madmom/ml/nn/layers.py

```diff
@@ -344,6 +344,144 @@ def activate(self, data):
         return out
 
 
+class GRUCell(object):
+    """
+    Cell as used by GRU layers proposed in [1].
+
+    Parameters
+    ----------
+    weights_in_to_hidden_update : numpy array, shape ()
+        Weights.
+    weights_in_to_hidden_update : numpy array, shape ()
+        Weights.
+    bias : scalar or numpy array, shape ()
+        Bias.
+    activation_fn : numpy ufunc, optional
+        Activation function.
+
+    References
+    ----------
+    .. [1] Cho, Kyunghyun, et al,
+           "On the properties of neural machine translation: Encoder-decoder
+           approaches",
+           arXiv preprint arXiv:1409.1259 (2014).
+
+    """
+
+    def __init__(self, weights_in_to_hidden_update,
+                 weights_hid_to_hidden_update, bias, activation_fn=tanh):
+        self.weights_in_to_hidden_update = weights_in_to_hidden_update
+        self.weights_hid_to_hidden_update = weights_hid_to_hidden_update
+        self.bias = bias
+        self.activation_fn = activation_fn
+
+    def activate(self, data, reset_gate, prev_hid):
+        """
+        Activate the gate with the given input, reset_gate and the previous
+        output:
+        h = tanh(Wxh * xt + Whh * ht-1 + b)
+
+        Parameters
+        ----------
+        data : scalar or numpy array, shape ()
+            Input data for the cell.
+        reset_gate : scalar or numpy array, shape ()
+            Activation of the reset gate.
+        prev_hid : scalar or numpy array, shape ()
+            Output data of the previous time step.
+
+        Returns
+        -------
+        numpy array
+            Activations of the gate for this data.
+
+        """
+        # weight input and add bias
+        in_to_hidden_update = np.dot(
+            data, self.weights_in_to_hidden_update) + self.bias
+        # weight previous hidden state and reset gate
+        hid_to_hidden_update = reset_gate * np.dot(
+            prev_hid, self.weights_hid_to_hidden_update)
+        # apply activation function and return it
+        return self.activation_fn(in_to_hidden_update + hid_to_hidden_update)
+
+
+class GRULayer(Layer):
+    """
+    Recurrent network layer with Gated Recurrent Units (GRU) as proposed in
+    [1].
+
+    Parameters
+    ----------
+    reset_gate : Cell instance
+        Reset gate.
```

Comment:
Although technically correct, it would be maybe clearer (and future proof) to use a `Gate` instead. `Cell`s are just there because they are present in old pickled models.

Sphinx renders nicer (with links and background etc.)  if it is written like this.
```
:class:`Gate`
```